src/app-layer-expectation.c

```diff
@@ -0,0 +1,287 @@
+    IPPairSetStorageById(ipp, expectation_id, exp);
+
+    SC_ATOMIC_ADD(expectation_count, 1);
+    IPPairUnlock(ipp);
+    return 0;
+
+error:
+    if (exp != NULL)
+        SCFree(exp);
+    return -1;
+}
+
+static Expectation *AppLayerExpectationGet(Flow *f, int direction, IPPair **ipp)
+{
+    Address ip_src, ip_dst;
+    if (GetFlowAddresses(f, &ip_src, &ip_dst) == -1)
+        return NULL;
+    if (direction & STREAM_TOSERVER) {
+        *ipp = IPPairLookupIPPairFromHash(&ip_src, &ip_dst);
+    } else {
+        *ipp = IPPairLookupIPPairFromHash(&ip_dst, &ip_src);
+    }
+    if (*ipp == NULL) {
+        return NULL;
+    }
+
+    return IPPairGetStorageById(*ipp, expectation_id);
+}
+
+/**
+ * Return Flow storage identifier corresponding to expectation data
+ *
+ * \return expectation data identifier
+ */
+int AppLayerExpectationGetDataId(void)
+{
+    return expectation_data_id;
+}
+
+static Expectation * RemoveExpectationAndGetNext(IPPair *ipp,
+                                Expectation *pexp, Expectation *exp,
+                                Expectation *lexp)
+{
+    SC_ATOMIC_SUB(expectation_count, 1);
+    if (pexp == NULL) {
+        IPPairSetStorageById(ipp, expectation_id, lexp);
+    } else {
+        pexp->next = lexp;
+    }
+    if (exp->data)
+        SCFree(exp->data);
+    SCFree(exp);
+    return lexp;
+}
+
+/**
+ * Function doing a lookup in expectation list
+ *
+ * \return an AppProto value if found
+ * \return ALPROTO_UNKNOWN if not found
+ */
+AppProto AppLayerExpectationLookup(Flow *f, int direction)
+{
+    AppProto alproto = ALPROTO_UNKNOWN;
+    IPPair *ipp = NULL;
+    Expectation *lexp = NULL;
+    Expectation *pexp = NULL;
+
+    int x = SC_ATOMIC_GET(expectation_count);
+    if (x == 0) {
+        return ALPROTO_UNKNOWN;
+    }
+
+    Expectation *exp = AppLayerExpectationGet(f, direction, &ipp);
+    time_t ctime = time(NULL);
+
+    if (exp == NULL)
+        goto out;
+
+    pexp = NULL;
+    while (exp) {
+        lexp = exp->next;
+        if ( (exp->direction & direction) &&
+             ((exp->sp == 0) || (exp->sp == f->sp)) &&
+             ((exp->dp == 0) || (exp->dp == f->dp))) {
+            alproto = exp->alproto;
+            if (FlowSetStorageById(f, expectation_data_id, exp->data) != 0) {
+                SCLogError(SC_ERR_UNKNOWN_VALUE, "Unable to set flow storage");
+            }
+            exp->data = NULL;
+            (void) IPPairDecrUsecnt(ipp);
+            /* remove the expectation */
+            exp = RemoveExpectationAndGetNext(ipp, pexp, exp, lexp);
+            continue;
+        }
+        /* Cleaning remove old entries */
+        if (exp && (ctime > exp->ts.tv_sec + EXPECTATION_TIMEOUT)) {
+            (void) IPPairDecrUsecnt(ipp);
+            /* remove the expectation */
+            exp = RemoveExpectationAndGetNext(ipp, pexp, exp, lexp);
+            continue;
+        }
+        pexp = exp;
+        exp = lexp;
+    }
+
+out:
+    if (ipp)
+        IPPairUnlock(ipp);
```

Comment:
think this should be IPPairRelease. It decrements the use_cnt that was increased during lookup.

Reply:
Good catch.